cdap-ranger/cdap-ranger-lookup/src/main/java/co/cask/cdap/security/authorization/ranger/lookup/client/CDAPRangerLookupClient.java

```diff
@@ -0,0 +1,222 @@
+   */
+  private ClientConfig getClientConfig() {
+    ClientConfig.Builder builder = new ClientConfig.Builder();
+    builder.setConnectionConfig(InstanceURIParser.DEFAULT.parse(
+      URI.create(instanceURL).toString()));
+
+    if (accessToken != null) {
+      builder.setAccessToken(accessToken);
+    }
+
+    String verifySSL = System.getProperty("verifySSL");
+    if (verifySSL != null) {
+      builder.setVerifySSLCert(Boolean.valueOf(verifySSL));
+    }
+
+    builder.setDefaultConnectTimeout(120000);
+    builder.setDefaultReadTimeout(120000);
+    builder.setUploadConnectTimeout(0);
+    builder.setUploadReadTimeout(0);
+
+    return builder.build();
+  }
+
+  /**
+   * Uses BasicAuthenticationClient to fetch {@link AccessToken}.
+   *
+   * @param username username to use while connecting
+   * @param password password for the above given username
+   * @return {@link AccessToken} for the user if authentication is successful
+   */
+  private AccessToken fetchAccessToken(String username, String password) throws IOException, TimeoutException,
+    InterruptedException {
+    Properties properties = new Properties();
+    properties.setProperty("security.auth.client.username", username);
+    properties.setProperty("security.auth.client.password", password);
+    final AuthenticationClient authClient = new BasicAuthenticationClient();
+    authClient.configure(properties);
+    ConnectionConfig connectionConfig = getClientConfig().getConnectionConfig();
+    authClient.setConnectionInfo(connectionConfig.getHostname(), connectionConfig.getPort(), false);
+    checkServicesWithRetry(new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        return authClient.getAccessToken() != null;
+      }
+    }, "Unable to connect to CDAP Authentication service to obtain access token, Connection info : " +
+      connectionConfig);
+    return authClient.getAccessToken();
+  }
+
+  /**
+   * Checks for system services to be up for SERVICE_TIMEOUT_SECONDS with 1 second delay during checks
+   */
+  private void checkServicesWithRetry(Callable<Boolean> callable,
+                                      String exceptionMessage) throws TimeoutException, InterruptedException {
+    long startingTime = System.currentTimeMillis();
+    do {
+      try {
+        if (callable.call()) {
+          return;
+        }
+      } catch (IOException e) {
+        // We want to suppress and retry on IOException
+      } catch (Throwable e) {
+        // Also suppress and retry if the root cause is IOException
+        Throwable rootCause = Throwables.getRootCause(e);
+        if (!(rootCause instanceof IOException)) {
+          // Throw if root cause is any other exception e.g. UnauthenticatedException
+          throw Throwables.propagate(rootCause);
+        }
+      }
+      TimeUnit.SECONDS.sleep(1);
+    } while (System.currentTimeMillis() <= startingTime + SERVICE_TIMEOUT_SECONDS * 1000);
```

Comment:
In that case it is too much. I think we should have connection timeout of 30 seconds, and total timeout of 90 or 120 seconds. What do you say?